### ISSUE TITLE
npm as a first-class task runner

### ISSUE THREAD

**Issue by boneskull** [npm/npm#11418]
*If this is a dupe, I apologize.  I'm always leery of adding to the pile of issues here.  However, I wanted to begin discussion around this (besides, it's more fun to think about than some `npm install` edge case).*

Many users are abandoning Gulp and Grunt for the npm `scripts` property.  I don't claim to know *all* of the reasons why this is happening, but I can gather a few things:

- Certain lifecycle scripts are not runnable w/o the development dependencies (ostensibly Grunt/Gulp) installed.  This means a user often has to put things in `scripts` *anyway*.  :+1: for keeping your scripts all in one place.
- Task runner plugins introduce bottlenecks to upgrades.  Introducing a dependency on Gulp or Grunt means you will also have dependencies on a variety of plugins.  If you wish to upgrade a package (say, `uglifyjs`), you may have to wait on the maintainer of `gulp-uglifyjs` or `grunt-uglifyjs` to ensure compatibility and update *that* package as well.
- npm is always present.  You do not need to add it as a dependency or development dependency.

On the main, npm works pretty well for this.  However, there are limitations to npm as a task runner (see #6943, #11291 and probably others).  For example, here's a (real) sample `scripts` property I'm working with:

```json
{
  "scripts": {
    "build": "parallel build:node build:browser",
    "build:node": "rimraf lib && babel src --out-dir lib",
    "build:browser": "rimraf dist && mkdirp dist && parallel build:browser:main build:browser:min",
    "build:browser:main": "browserify --transform [ exposify --expose [ --angular angular ] ] --plugin [ bannerify --file .banner.ejs --pkg package.json ] src --outfile dist/angular-loggly-mixin.js",
    "build:browser:min": "browserify --transform [ exposify --expose [ --angular angular ] ] --plugin [ bannerify --file .banner.ejs --pkg package.json ] --global-transform uglifyify src --outfile dist/angular-loggly-mixin.min.js",
    "bump:major": "npm version major -m 'Release v%s'",
    "bump:minor": "npm version minor -m 'Release v%s'",
    "bump:patch": "npm version patch -m 'Release v%s'",
    "check": "parallel nsp deps",
    "deps": "parallel deps:missing deps:extra",
    "deps:extra": "dependency-check package.json --no-default-entries --entry src/index.js --extra --no-dev",
    "deps:missing": "dependency-check package.json --no-default-entries --entry src/index.js --missing",
    "deps:update": "updtr",
    "lint": "eslint src/ test/ .*.js",
    "nsp": "nsp check",
    "prepublish": "npm run build",
    "preversion": "npm run test",
    "publish-please": "publish-please",
    "test": "parallel check lint test:unit test:e2e",
    "test:unit": "karma start .karma.unit.conf.js --single-run",
    "test:e2e": "npm run build:browser && karma start .karma.e2e.conf.js --single-run"
  }
}
```

:weary: There are some features I can suggest given the above:

1. #6943 (depending on implementation) would help make things more clear.  If it was assumed an `Array` implied `&&`, then `build:node` could be rewritten as:

  ```json
  {
     "build:node": [
       "rimraf lib",
       "babel src --out-dir lib"
     ]
  }
  ```

  `--force` could be leveraged to imply `;` instead of `&&`.
2. Hierarchical scripts.  I have spent a few hours trying to hack a solution together, but I can't do it without either abusing `npmrc` or installing an extra dependency.  When coupled with the previous bullet, this would mean `build` could become:

  ```json
  {
    "build": {
      "node": [
        "rimraf lib", 
        "babel src --out-dir lib"
      ],
      "browser": [
        "rimraf dist", 
        "mkdirp dist",
        {
          "main": "browserify --transform [ exposify --expose [ --angular angular ] ] --plugin [ bannerify --file .banner.ejs --pkg package.json ] src --outfile dist/angular-loggly-mixin.js",
          "min": "browserify --transform [ exposify --expose [ --angular angular ] ] --plugin [ bannerify --file .banner.ejs --pkg package.json ] --global-transform uglifyify src --outfile dist/angular-loggly-mixin.min.js"
        }
      ]
    }
  }
  ```

  Given the above, `npm run build.browser` would `rimraf`, `mkdirp`, then execute the `main` target and `min` target.  It would *not* run `rimraf` and `mkdirp` *before each* target; if this was necessary, the array format could be used instead:

  ```json
  {
    "browser": {
      "main": [
        "rimraf dist", 
        "mkdirp dist",
        "browserify --transform [ exposify --expose [ --angular angular ] ] --plugin [ bannerify --file .banner.ejs --pkg package.json ] src --outfile dist/angular-loggly-mixin.js"
      ],
      "min": [
        "rimraf dist", 
        "mkdirp dist",
        "browserify --transform [ exposify --expose [ --angular angular ] ] --plugin [ bannerify --file .banner.ejs --pkg package.json ] --global-transform uglifyify src --outfile dist/angular-loggly-mixin.min.js"
      ]
    }
  }
  ```

  Removing `dist/` before each wouldn't make sense; it's just an example.
3. References to other scripts.  To eliminate duplication of `rimraf` and `mkdirp`, you can currently nest a call to `npm run-script`:

  ```json
  {
    "reset-dist": [
      "rimraf dist", 
      "mkdirp dist"
    ],
    "browser": {
      "main": [
        "npm run reset-dist",
        "browserify --transform [ exposify --expose [ --angular angular ] ] --plugin [ bannerify --file .banner.ejs --pkg package.json ] src --outfile dist/angular-loggly-mixin.js"
      ],
      "min": [
        "npm run reset-dist",
        "browserify --transform [ exposify --expose [ --angular angular ] ] --plugin [ bannerify --file .banner.ejs --pkg package.json ] --global-transform uglifyify src --outfile dist/angular-loggly-mixin.min.js"
      ]
    }
  }
  ```

  That's workable. Another solution would be to provide "shortcuts"--something like `@reset-dist`.  But that introduces a weird, arbitrary syntax that we likely wish to avoid.  Given we're limited to JSON, I'm not sure it's feasible. 
4. Support `package.yaml`.  We would then be able to do all sorts of crazy stuff like the above.
5. Parallel script execution
6. Pie-in-the-sky would be to provide some cross-platform utilities (like `rimraf`, `mkdirp`, `realpath`, etc) out-of-the-box.  Given that `rimraf` and `mkdirp` are *dependencies* of npm, maybe just expose them to the CLI?

npm should *avoid* "plugins" to support various behaviors, or we end up with bottlenecks, which defeats the purpose of all of this. Plugins aren't necessary.  

With just a couple improvements, npm could go from "adequate" to "awesome" at running scripts.

Please let me know what you think.

**Comment by iarna** [comment]
This suffers from the same problem described in #11284 in that `scripts` are not just a task runner, they're also used as a part of `npm`'s plumbing, and allowing different formats in there would lock out users of older versions of npm, which isn't something we're keen on doing.

Changing `package.json` to YAML has similar problems, although we probably could work around those. Still, you would have a module that during development Node.js doesn't know how to load and that makes it a really hard sell.

The cli team itself is scheduled out for the coming year and I don't see any place this would fit, even if it didn't have the issues I've discussed. As such, I'm gonna close this this issue. This doesn't actually preclude us taking a PR for work, but in this case, breaking older npm's means this solution would need some more work before we could accept it.

Personally, I would approach this as an external task runner that would differ from grunt and gulp by not having plugins and just executing scripts found in the `package.json` or other local config. You are free to add whatever new keys to `package.json` you want– npm only validates the keys it uses itself.

**Comment by boneskull** [comment]
> allowing different formats in there would lock out users of older versions of npm

There are a few things that mitigate this:

1.  Tools like `nvm`
2.  The `engine` field
3.  Package authors don't release a minor or patch version which depends on the new functionality

It's painful to break an API.  But it's also tough to make an omelette w/o cracking a few eggs. :smile:

> Still, you would have a module that during development Node.js doesn't know how to load and that makes it a really hard sell.

Can you explain this further?  npm has many userland dependencies.

> Personally, I would approach this as an external task runner that would differ from grunt and gulp by not having plugins and just executing scripts found in the package.json or other local config. You are free to add whatever new keys to package.json you want– npm only validates the keys it uses itself.

That only solves one of the issues.  

- This task runner would still have to be a (production) dependency if it was to be used with, say, `postinstall`
- Nobody wants to install `grunt` as a production dependency (nor should they); same goes for this
- The task runner can't leverage `prepublish` functionality without either using `scripts` or abusing `hooks`

> This suffers from the same problem described in #11284 in that scripts are not just a task runner, they're also used as a part of npm's plumbing,

Yes!  Access to its plumbing is vital.  That's why I made the issue here instead of writing another task runner.  #11284 is a great idea in general.

You'll likely see more & more tickets like this one and #11284 going forward.  If you already have a roadmap for the next year, then it's simply not a priority right now.

I'm happy resuming discussion at a later date.  I do not see any way to address these types of issues without breaking the API, which means it's going to need a lot more attention and discussion when the time comes.  What's the best way to advocate for it?

**Comment by boneskull** [comment]
(I can't just whip up a PR if I'm going to break the API w/ it; don't want to waste my time or anyone else's)

**Comment by iarna** [comment]
(This is still open because we're going to have a bit of further discussion internally.)

**Comment by iarna** [comment]
A lot of what I have to say is predicated on this:

An awful lot of folks are stuck in 0.10 and 0.12 still and even fewer actually upgrade their copy of `npm`. Often this is due to it department decrees that the developers have no control over. Not that there's anything special about those versions per se, but the point being that many organizations are very resistant to updating Node.js let alone `npm`. We want to support these users too.

This means that we have to pay a lot of attention to what I call forward compatibility. That is, artifacts produced by newer versions of npm shouldn't break users of older npm's. Breaking backward compatibility is just fine (in a major release), because you have to opt into it by upgrade to the new major release. But I at least am going to be very cautious about any change that breaks forward compatibility.

> There are a few things that mitigate this:
>  1. Tools like nvm

I love `nvm` and use it all the time, but I'm not clear on how you think this helps mitigate anything? It helps manage different `node` versions for developers, but it doesn't help with convincing folks to update their `npm`, either locally or in production.

> The engine field

The engine field does issue a warning if your engine doesn't match what was requested and so at least it would be clear that the installed module might be totally broken– for the users I'm talking about here, it would give them the opportunity to switch to a different module or version of the module.  But it's little comfort to folks when it's a dependency of a dependency of a dependency that's issuing this. Fixing _that_ means using `npm shrinkwrap` in ways that are still awkward to pin the version of that one dep in particular.

> Package authors don't release a minor or patch version which depends on the new functionality

Ideally yes, but say a new module is published to the registry and I decide to bring it into my project. I would have to then audit all of its package scripts (and all of it's dependencies' package scripts, and it's dependencies' dependencies' package scripts) to see if they needed the newer npm and thus called for a major version bump to my own module. In practice, I just can't see folks doing that, and so I would envision this introducing breaking changes regularly.

> > Still, you would have a module that during development Node.js doesn't know how to load and that makes it a really hard sell.

So the one use of `package.json` that Node.js makes is to look up the name of the main entry point script– it looks for a `main` entry and if that's missing it uses `index.js`. If you used a `package.yaml` then 
Node.js has no way to do that and so if you use anything other than `index.js`. The work around for that is to generate a `package.json` at some point, but then that can get out of sync. If you want to switch to something that's not json, imo, your first stop is the Node.js project, but I wouldn't expect them to be very receptive. But if they did go for it then `npm` would certainly support it one way or another.

> That only solves one of the issues.
> 
> * Nobody wants to install grunt as a production dependency (nor should they); same goes for this
If it's light enough weight folks would be less opposed I think.
> * The task runner can't leverage prepublish functionality without either using scripts or abusing hooks

Personally, I think there is value in distinguishing between things that are tied to the opperation of npm itself (all of the various lifecycle hooks) and things that are tasks for humans to run. But that said, I probably _would_ include the ability to have the task runner add calls to itself to the scripts section automatically.

**Comment by boneskull** [comment]
I'll respond further later, but:

> Ideally yes, but say a new module is published to the registry and I decide to bring it into my project. I would have to then audit all of its package scripts (and all of it's dependencies' package scripts, and it's dependencies' dependencies' package scripts) to see if they needed the newer npm and thus called for a major version bump to my own module. In practice, I just can't see folks doing that, and so I would envision this introducing breaking changes regularly.

You make it sound more difficult than it is.  :smile: Use `nvm` to downgrade and attempt to install the new package.  Anyway, I live with this reality wherein I maintain a package which must support npm v1.2.30.  I can't use any version of any package with a caret in its `dependencies` field.  

The message I'm getting is "we can't *ever* modify the behavior of any field in `package.json` because it will break things."  I'm not sure if this is your intent, but it's discouraging.  Package maintainers who don't RTFM and carelessly use a breaking npm feature should not prevent npm from moving forward.

> Personally, I think there is value in distinguishing between things that are tied to the opperation of npm itself (all of the various lifecycle hooks) and things that are tasks for humans to run. But that said, I probably would include the ability to have the task runner add calls to itself to the scripts section automatically.

This is a good point (and one which I actually wrote up myself, then deleted).  Perhaps we could achieve the bulk of this functionality in a new field for scripts which are not bound to lifecycle events.  That'd get us at least part way to the goal.

**Comment by othiym23** [comment]
Thanks for the thorough writeup and for talking this through with @iarna, @boneskull. While the CLI team isn't going to ship this particular set of solutions, you've gone a great job of clarifying what people are using package scripts for, and raised some good points about problems with the current implementation. I'll put down my thoughts here, and enumerate what I think are some approaches worth investigating for addressing those problems.

The current popularity of package scripts was more or less unplanned. Most of that functionality is a byproduct of npm's primary function as a package manager. Package scripts themselves started as a simple extension to the npm package lifecycle runner, and `npm run-script` has been the beneficiary of a bunch of further tweaks since then. Over the last year or so, people have noticed that an awful lot of common build and maintenance tasks can be done purely with `devDependencies` and fairly simple package scripts, to the point that they're now being considered a viable alternative to dedicated build tools like Gulp, Grunt, and Broccoli. This is great!

However, I think it's worth spending a little time thinking about where that popularity comes from. At least part of it seems to be because the current implementation hits a sweet spot between simplicity and power. The fact that it's very difficult to come up with a build pipeline as complex as one you could create in a Makefile or Rakefile, or define with a custom Gulp pipeline, is a feature as well as a limitation. The snippets are easy to understand and copy, the "plugin API" naturally becomes binaries exported by packages that can be plugged into `devDependencies`, and when things start getting unwieldy, that's a pretty good signal that you should rethink your approach.

At the same time, there are going to remain a decent-sized class of use cases that are more naturally handled by a dedicated tool like Gulp. To try to pull all of that power into npm itself would come at the cost of making a fairly simple feature much less teachable (and, in all likelihood, supportable), and it seems at least unnecessary to do that when tools like Gulp and Grunt already exist and have large and healthy communities of their own.

With that out of the way, let me discuss your specific proposals:

1. As @iarna wrote, the technical problem with using arrays and nested object literals to decompose package scripts is that old versions of npm aren't forward-compatible with either, meaning this is a breaking change to the ecosystem. However, as important is the product problem this poses: it introduces a significant amount of cognitive overhead for people learning how to use the feature. Almost every instance where npm supports different types for fields in `package.son`, it's turned out to be a mistake, as both the maintenance and support costs turn out to be significant. This also makes it much easier for developers to come up with tangled rats' nests of interrelated tasks that I, at least, am all too familiar with from looking at the `Makefile`s, `build.xml`s, and `Rakefile`s of large projects. I think this is an especially significant concern for npm because `package.json` is already so big and unwieldy. Pressure to keep the scripts section relatively terse is valuable, in my opinion. I think this last point also covers the case of putting scripts in a different field with another name.
2. There have been many, many requests to use something other than JSON for `package.json`, but that ship sailed years ago. My current preference is to make it easier to use npm itself to manage the contents of `package.json`, and to make it easier to use those features than to edit the file by hand. Comments would be nice. However, this is another instance of changes breaking every version of npm older than the version that can handle those changes. That said, there are forks of npm out there that do use YAML files (see: [`yapm`](http://npm.im/yapm), and I don't think you'd need a fully-fledged package manager to handle translating between JSON and more human-readable configuration file formats. This isn't something that needs to be in npm proper.
3. Shortcuts require npm to use something other than a command shell to interpret the comments of package scripts, and if you're going to be parsing the scripts (which npm currently doesn't need to do), you might as well go for broke (I have more to say about this in a moment).
4. Parallel script execution, especially on Windows, is indeed tricky right now. It can be done using [`parallel`](http://npm.im/parallel) or similar utilities, but it would be very useful if npm provided a simple version of task concurrency itself. (I also have more to say about this further down.)

> Package maintainers who don't RTFM and carelessly use a breaking npm feature should not prevent npm from moving forward.

I agree with your sentiments, but I think that the maintainers of npm itself should probably hold themselves to a higher standard. 😅 I don't say this to be glib; we have and will continue to change the meaning of fields in `package.json` when it makes sense, but it requires a very high level of care in an ecosystem with 200K+ packages. For an example, we changed the type of the `"license"` field from one kind of a string to another, and that ended up consuming a startling amount of the team's time in the second half of last year.

> Anyway, I live with this reality wherein I maintain a package which must support npm v1.2.30.  I can't use any version of any package with a caret in its `dependencies` field. 

The rollout of the caret operator is a cautionary tale for me. It's caused tension between npm and several of Node's core maintainers (because of strong differences of opinion about whether `^` or `~` is safer, and because of it being a backwards-incompatible change that breaks working 0.8 installs. It's a big part of the reason why npm still needs to support Node 0.8.

> What's the best way to advocate for it?

Small changes submitted in rough PRs. Even if they're potentially semver-major changes, they should be submitted as discrete units of work, and it's in everyone's best interests to keep the various changes as decoupled as possible.

That last response points out that I do think there's room for innovation here, although the CLI team isn't going to have the time to get to it in the next 6-12 months, in all likelihood. So what do I think might be useful? I think you were onto something with this:

> Pie-in-the-sky would be to provide some cross-platform utilities (like `rimraf`, `mkdirp`, `realpath`, etc) out-of-the-box.

See #11291. We've been mulling over the idea of codifying some subset of shell scripting that we'd expect to see supported in a cross-platform way, and then writing a parser that takes that and wraps it around something like [`shelljs`](http://npm.im/shelljs). This could also include the idea of backgrounding tasks or executing tasks in parallel (although I should make clear here that I'm no more interested in making npm into a heavyweight alternative for nodemon than I am an alternative to Gulp).

Another thing I've been thinking about is crafting a DSL designed to handle the most common set of use cases encompassed by the package scripts out in the wild today. My tendency would be to start as minimally as possible and build up very gradually from there. It would be necessary use some syntactic marker (like `##!`, maybe?) that would at worst make these scripts a no-op in old versions of the CLI.

@iarna has already raised most of the points that make this fraught, and there are even more out-there possibilities that we've already rejected for now, like #11284. That said, if people wanted to play around with pieces of the above, it would be really interesting to hear how it goes.

The team is open to changes that make package scripts more powerful and useful, but it's also very important to me that we don't inadvertently wreck a feature the community has adopted and made its own. Thanks again for going so deep on this with us, and I hope this explanation makes sense to you.

**Comment by boneskull** [comment]
@othiym23 Thanks for taking the time to type that up.

I'm hopeful we can make incremental improvements here.  I'll take a look at #11291; I think leveraging `shelljs` would be a cool idea, but we can kill a flock with one stone w/o introducing a DSL.  (I'll comment there)

The point about keeping things easy is a good one.  That being said, I understand now supporting some sort of hierarchical data structure for `scripts` and its consequences is not worth it.

(Yes, my example `scripts` prop is unwieldy.  Grunt/Gulp/etc are (arguably) better tools for this sort of job.  I'm just very stubborn.)
